pkg/aspect/types.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aspect
+
+import (
+	"io"
+
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/genproto/googleapis/rpc/code"
+
+	"istio.io/mixer/pkg/attribute"
+	"istio.io/mixer/pkg/expr"
+
+	istioconfig "istio.io/api/istio/config/v1"
+)
+
+type (
+
+	// Aspect -- User visible cross cutting concern
+	// istio/metrics, istio/accessControl, istio/accessLog are all aspects of a deployment
+	// This interface is extended by aspects;
+	// ex: listChecker.Aspect adds --> CheckList(Symbol string) (bool, error)
+	// The extended interface is implemented by adapter impls.
+	// ex: //adapter/ipListChecker aspectState implements the listChecker.Aspect.
+	Aspect interface {
+		io.Closer
+		// ImplName returns the official name of adapter implementation ex. "istio/statsd"
+		ImplName() string
+	}
+
+	// Config (aspect.Config) is the internal struct used for the Aspect proto
+	// at present it only has an additional field that stores the converted
+	// and typed proto
+	Config struct {
+		istioconfig.Aspect
```

Comment:
This remains very confusing.

This type is called Config and its first member is something called an Aspect. This is not related to the type Aspect right above though. One Aspect is an interface with methods, another Aspect is a block of config.